### ISSUE TITLE
Make setting identifiers easier to use

### ISSUE THREAD

**Issue by jstewmc** [jstewmc/gravity#6]
We should make setting identifiers easier to use. While _service identifiers_ have PHP's convenient `::class` constant, _setting identifiers_ are manual strings. 

One of our initial ideas was to use a prefix or a placeholder like PHP's `namespace` declaration to prevent typing the same prefix over and over again. 

For example, you could define the prefix at the top of the file (`in` is just a back-of-the-envelope method name):

```php
// set the placeholder to set a configuration "domain"
$g->in('foo.bar.baz');

// use the placeholder (this would be equivalent to 'foo.bar.baz.qux') (is the `$` the best character?)
$g->set('$.qux'); 

// call this explicitly at the end of the file to exit the domain
$g->out();
```
Gravity should automaticaly call the `out()` method between loading files in order to prevent side effects.

The `$` placeholder seems like a risky character since it prefixes variables in PHP, and it would make double quotes harder to use. However, it's a common placeholder in other languages like Javascript. 

I don't think Gravity should try to implicitly set the placeholder before the file is loaded, because that could have side effects as well. A developer should explicitly define it at the top of the file like PHP's `namespace` declaration.

**Comment by jstewmc** [comment]
Hmm, rather than an opt-in placeholder like suggested above, which is the opposite of PHP's `namespace` opt-out system, why don't we do a setting identifier opt-out? It would follow PHP's namespace logic (i.e., most things defined within a namespace live there). 

We could even use the methods `namespace` and `use`. According to the documentation on [PHP keywords](http://php.net/manual/en/reserved.keywords.php):

> As of PHP 7.0.0 these keywords (e.g., use, namespace, etc) are allowed as property, constant, and method names of classes, interfaces and traits, except that class may not be used as constant name.

I don't want to replace the `namespace`, `use`, and `::class` system for service identifiers, but an equivalent system for setting identifiers might be nice:

```php 
$g->namespace('jstewmc.gravity.example');

$g->set('foo.bar.baz');   // sets "jstewmc.gravity.example.foo.bar.baz" 

// we could use leading-separators as namespace exits
$g->set('.foo.bar.baz');  // sets "foo.bar.baz"

// we could support use statements too
$g->use('jstewmc.gravity.example.qux');

$g->set('qux');       // sets "jstewmc.gravity.example.qux"
$g->set('qux.quux');  // sets "jstewmc.gravity.example.qux.quux"

// and support aliases for use statements
$g->use('jstewmc.gravity.example.quux', 'home');

$g->set('home');  // sets "jstewmc.gravity.example.quux" 
```

We could support service namespaces and settings if we wanted. There isn't a huge need, though. PHP's `namespace`, `use`, and `::class` system is pretty convenient.

If we don't support services as well, we'd probably need to add some sort of "c" or "config" prefix to the method names (e.g., "cnamespace") to make it clear that these only work for configuration, which I'm not dying to do.

**Comment by jstewmc** [comment]
@javabudd, what do you think about creating an opt-out `namespace` and `use` system for setting identifiers, rather than an `opt-in` placeholder system?

**Comment by jstewmc** [comment]
Keep in mind, PHP's `::class` constant doesn't include the leading-separator. 

But, in the same way we determine an identifier's type by the separator, we could determine the type of the `use` and `namespace` values. 

Honestly, we could probably treat the `namespace` and `use` values as identifiers, if we drop the requirement that identifiers contain three segments, which is frankly, a holdover from when we used vendors and packages to order things.

**Comment by jack-clayton** [comment]
Hmm, we probably only want to support the `namespace` and `use` methods in a file, where we can clear the settings after the file has loaded. Otherwise, it would be up to the user to remember to clear/reset/exit the namespace and use statements at the right time, which seems unexpected and could produce major side effects. Most of Gravity's settings and services should be defined in files anyway. The ability to do things outside of files is really just for the rare convenience. 

If you want to use a "namespace" outside of a file, just use string interpolation:

```php
$namespace = 'foo.bar.baz';

$g->set("$namespace.qux.quux", 1);
```

This will likely require changes to the _load-filesystem_ service. It could wrap the manager and provide its own private `get()`, `set()`, `alias()`, and `deprecate()` methods. It could pre-proces the identifiers based on the file's "namespace" and "use statements" before handing the identifiers and data to the manager.

**Comment by jstewmc** [comment]
To support this feature, I'm moving the identifier's logic to a new object, `Path`. A path doesn't have to be an end-to-end identifier. It may just be a partial location, unlike an identifier which must contain at least three segments.

**Comment by jstewmc** [comment]
Dang. We're probably going to have to pass a file's context to anonymous function defintions. That way, when they are instantiated, the paths inside the function can be resolved using the file's rules. Otherwise, the definition would use the file's context but the anonymous function wouldn't. I imagine that would be pretty unexpected behavior. 

We should make sure the `namespace()` method is the first function call in the Gravity file, and we should make sure any `use()` calls occur before other method calls. Otherwise, the file's context would change as the file was read, which seems like a bad idea.

**Comment by jstewmc** [comment]
Closed by #34.
